### PR TITLE
[Single] invoke with pre-allocated output data

### PR DESCRIPTION
**Pull request by jaeyun-jung**
Add new function to invoke the model with pre-allocated output data handle.
This may reduce single-shot invoke latency in Android Java API.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

**Comment by jaeyun-jung** [comment]
Test on S20+ (QC)
model: tensorflow-lite ssd

- Before this PR (output data conversion)
2020-09-22 17:32:16.491 21844-21881/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] count 100
2020-09-22 17:32:16.491 21844-21881/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] prepare 40
2020-09-22 17:32:16.491 21844-21881/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] invoke 210755
2020-09-22 17:32:16.492 21844-21881/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] result 1047
2020-09-22 17:32:16.492 21844-21881/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] total 211843

- Use direct buffer address
2020-09-22 20:44:53.113 20055-20111/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] count 100
2020-09-22 20:44:53.113 20055-20111/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] prepare 43
2020-09-22 20:44:53.113 20055-20111/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] invoke 222486
2020-09-22 20:44:53.113 20055-20111/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] result 111
2020-09-22 20:44:53.113 20055-20111/org.nnsuite.nnstreamer.test W/nnstreamer: [PROFILE] total 222640

File: api/capi/src/tensor_filter_single.c [hunk 1]
```diff
@@ -299,25 +303,49 @@ g_tensor_filter_single_invoke (GTensorFilterSingle * self,
...
 
-  if (self->allocate_in_invoke == FALSE) {
+  /* set output tensors for given params */
+  _out = output;
+
+  if (self->allocate_in_invoke) {
+    if (!allocate) {
+      /**
+       * @todo how can we remove memcpy if output data is already allocated
+       * single-shot should fill the output data, but sub-plugin allocates new memory.
```
Comment (myungjoo) [inline api/capi/src/tensor_filter_single.c:313]: If the filter allocates in invoke unconditionally, there is no way to skip memcpy if allocate==TRUE :)

**Review by myungjoo — APPROVED** [review]
LGTM

**Review by again4you — APPROVED** [review]
LGTM!

File: api/android/api/src/main/jni/nnstreamer-native-singleshot.c [hunk 2]
```diff
@@ -227,24 +228,35 @@ nns_native_single_invoke (JNIEnv * env, jobject thiz, jlong handle, jobject in)
...
+  if (ml_single_invoke_no_alloc (single, in_data, out_data) != ML_ERROR_NONE) {
+    nns_loge ("Failed to invoke the model.");
+    failed = TRUE;
+    goto done;
   }
 
 done:
-  /* do not free input tensors (direct access from object) */
+  if (failed) {
+    if (result) {
```
Comment (again4you) [inline api/android/api/src/main/jni/nnstreamer-native-singleshot.c:251]: What about reducing just one line?
```cpp
if (failed && result) {
```

**Review by again4you — APPROVED** [review]
LGTM!

**Review by dongju-chae — APPROVED** [review]
LGTM

**Review by kparichay — APPROVED** [review]
LGTM

File: api/android/api/src/main/jni/nnstreamer-native-singleshot.c (hunk 2 above)
Comment (myungjoo) [reply api/android/api/src/main/jni/nnstreamer-native-singleshot.c:251]: @again4you Please remember and apply it after the release or at for-next :)

**Pull request merged**